server/src/main/java/org/apache/druid/segment/realtime/appenderator/StreamAppenderator.java

```diff
@@ -337,6 +337,9 @@ public AppenderatorAddResult add(
       throw e;
     }
 
+    final long currTs = System.currentTimeMillis();
+    metrics.reportMessageGap(currTs - row.getTimestampFromEpoch());
```

Comment:
We should just send the message timestamp here. `SegmentGenerationMetrics` can decide what to do with it, same way we currently do for the max message timestamp.
```suggestion
    metrics.reportMessageTimestamp(row.getTimestampFromEpoch());
```

Reply:
I'd argue against this if we care about performance. This function already does a clock read in a hot-loop. Doing another call just slows down the function even more.